aten/src/ATen/new_gen.py

```diff
@@ -0,0 +1,953 @@
+            'extra_cuda_headers': '',  # TODO: remove this
+            'type_derived_method_declarations': compute_type_derived_method_declarations(dispatch),
+        }
+        f.write(TYPE_DERIVED_H.substitute(env))
+
+JIT_TO_CPP_DEFAULT = {
+    'False': 'false',
+    'True': 'true',
+    'None': 'c10::nullopt',  # UGH this one is type directed
+    'Mean': 'at::Reduction::Mean',
+    '[]': '{}',
+    '[0,1]': '{0,1}', # TODO: stop special casing
+    'contiguous_format': 'MemoryFormat::Contiguous',
+}
+
+def cpp_default(d: str, t: Type) -> str:
+    if d == 'None' and str(t) == 'Tensor?':
+        return '{}'
+    return JIT_TO_CPP_DEFAULT.get(d, d)
+
+def compute_function_declarations() -> List[str]:
+    rs: List[str] = []
+    for f in native_functions:
+        with context(f'in {f.loc}:\n  {f.func}'):
+            if f.manual_kernel_registration:
+                continue
+            if Variant.function not in f.variants:
+                continue
+
+            # TODO: clear up naming
+            cpp_return = cpp_type_return(f.func.returns)
+            name = str(f.func.name.name)
+            if f.func.is_out_fn():
+                name += '_out'
+
+            # TODO: this was copy pasted
+            def format_arg(a: Argument) -> str:
+                # DEFAULTING IS NEW
+                default = f"={cpp_default(a.default, a.type)}" if a.default is not None else ""
+                # TODO: Always NO legacy optional
+                return f"{cpp_type(a.type, mutable=a.is_write, argument=True, legacy_optional=not f.use_c10_dispatcher_full)} {a.name}{default}"
+            cpp_args: List[str] = []
+            cpp_args.extend(map(format_arg, f.func.out_arguments))
+            cpp_args.extend(map(format_arg, f.func.arguments))
+
+            # Discover TensorOptions
+            topt_names = ['dtype', 'layout', 'device', 'pin_memory']
+            kwargs = list(f.func.kwarg_only_arguments)  # short name
+            i = 0
+            while i < len(kwargs):
+                if i <= len(kwargs) - len(topt_names) and all(kwargs[i+j].name == topt_names[j] for j in range(len(topt_names))):
+                    if str(f.func.name) in ["_cudnn_init_dropout_state", "sparse_coo_tensor.size", "_sparse_coo_tensor_with_dims", "_sparse_coo_tensor_with_dims_and_tensors"]:
+                        # I think this is a bug in the original
+                        cpp_args.append('const TensorOptions & options')
+                    elif str(f.func.name) in ["tril_indices", "triu_indices"]:
+                        cpp_args.append('const TensorOptions & options=at::kLong')
+                    else:
+                        cpp_args.append('const TensorOptions & options={}')  # MODIFIED
+                    i += len(topt_names)
+                else:
+                    cpp_args.append(format_arg(kwargs[i]))
+                    i += 1
+
+            rs.append(f"CAFFE2_API {cpp_return} {name}({', '.join(cpp_args)});")
+    return rs
```

Comment:
ok I know this is way more than you were looking for but per our chat this afternoon here's a version of `compute_function_declarations()` that uses some moves that kept me sane when I rewrote `gen_python_functions` way back when. 

* ruthlessly pulling functions up to the top level and plumbing all their dependencies into them as explicit params. The original `compute_function_declarations()` is still pretty small and clear, but reading the dependencies between local functions and closed-over locals is already getting nontrivial... whereas in the flattened version it's super obvious what e.g. formatting an arg depends on, what formatting a tensor options arg depends on (just the function name!) etc.

* analogously, yanking any data constants up into global space and throw all available info into their names and comments. Helps visibility and avoiding drift-prone repetition of special cases buried inside functions... and for the many, many of these that are pattern data for workarounds, it demarcates them more clearly as the eruptions of irregularity that they are 😁 

* being super finicky about missed patterns (here yelling about out-of-sequence tensor options args)

I went ahead and did the whole exercise here just to show how the end result helps (me, anyway) distinguish between the general case and the special cases, as well as what input is being fed into each decision. Well, plus I did it bc I got carried away :P 

One last pitch - it might seem like overkill to be so hardcore, now while the code is pretty tight and tractable - but by the time I got to `gen_python_functions` some of the analogs to `compute_function_declarations()` were like multi-hundred line behemoths with nested and doubly-nested inner functions that just closed over everything, big dictionaries getting tweaked miles and miles away, the whole deal. I think the flatter and more explicitly plumbed we make things now, the easier it'll be to fight that kind of entropy.

```
# format an argument for c++
def format_arg(a: Argument, legacy_optional: bool) -> str:
    # DEFAULTING IS NEW
    default = f"={cpp_default(a.default, a.type)}" if a.default is not None else ""
    # TODO: Always NO legacy optional
    return f"{cpp_type(a.type, mutable=a.is_write, argument=True, legacy_optional=legacy_optional)} {a.name}{default}"

# tensor options attributes come in scattered, we recognize and convert to a single TensorOptions arg
TOPT_NAMES = ['dtype', 'layout', 'device', 'pin_memory']
TOPT_LEN = len(TOPT_NAMES)

# true if i begins the scattered tensor options arguments in args.
# note: it's an error for these to be out of order. (TODO what about noncontiguous?)
def at_topt_args(i: int, args: List[Argument]) -> bool:
    in_seq = i <= len(args) - TOPT_LEN and all(args[i+j].name == TOPT_NAMES[j] for j in range(TOPT_LEN))
    perm = i <= len(args) - TOPT_LEN and set([a.name for a in args[i:i + TOPT_LEN]]) == set(TOPT_NAMES)
    if in_seq != perm:
        raise ValueError(f"TensorOptions arguments must be specified in the following order: {', '.join(TOPT_NAMES)}")
    return in_seq

# TODO these need TensorOptions without a default for some reason
TENSOR_OPTIONS_NO_DEFAULT = [
    "_cudnn_init_dropout_state",
    "sparse_coo_tensor.size",
    "_sparse_coo_tensor_with_dims",
    "_sparse_coo_tensor_with_dims_and_tensors"]

# TODO these need TensorOptions defaulted to (dtype) long
TENSOR_OPTIONS_DEFAULT_LONG = ["tril_indices", "triu_indices"]

# format TensorOptions arg. Handle some special cases by name
def format_topt_arg(f: NativeFunction):
    if str(f.func.name) in TENSOR_OPTIONS_NO_DEFAULT:
        # I think this is a bug in the original
        return 'const TensorOptions & options'
    elif str(f.func.name) in TENSOR_OPTIONS_DEFAULT_LONG:
        return 'const TensorOptions & options=at::kLong'
    else:
        return 'const TensorOptions & options={}'  # MODIFIED

# compute all c++ function declarations 
def compute_function_declarations() -> List[str]:
    rs: List[str] = []
    for f in native_functions:
        with context(f'in {f.loc}:\n  {f.func}'):
            if f.manual_kernel_registration:
                continue
            if Variant.function not in f.variants:
                continue

            # TODO: clear up naming
            cpp_return = cpp_type_return(f.func.returns)
            name = str(f.func.name.name)
            if f.func.is_out_fn():
                name += '_out'

            cpp_args: List[str] = []
            cpp_args.extend(map(lambda a: format_arg(a, not f.use_c10_dispatcher_full), f.func.out_arguments))
            cpp_args.extend(map(lambda a: format_arg(a, not f.use_c10_dispatcher_full), f.func.arguments))

            # Discover TensorOptions
            kwargs = list(f.func.kwarg_only_arguments)  # short name
            i = 0
            while i < len(kwargs):
                if at_topt_args(i, kwargs):
                    cpp_args.append(format_topt_arg(f))
                    i += len(TOPT_NAMES)
                else:
                    cpp_args.append(format_arg(kwargs[i], not f.use_c10_dispatcher_full))
                    i += 1

            rs.append(f"CAFFE2_API {cpp_return} {name}({', '.join(cpp_args)});")
    return rs
```

Reply:
Yes, pulling things up seems appropriate. One challenge is I need some sort of top level organizational principle (you toss things as local definitions so you don't have to worry about this.)

Reply:
For ordering I generally try to stick to in-order, with comment headers to replicate the visual effect of nesting where it helps. So e.g. the code block above might have a big 
```
# --------- compute_function_declarations and helpers -----------
#
# ...
#
``` 
at the top, and maybe a smaller one above the stuff about TensorOptions args.

You're right that it forgoes the natural structuring effect of nested functions, which is a bummer. But the readability payoff of explicit dependencies is definitely worth I think